packages/block-editor/src/hooks/block-bindings.js

```diff
@@ -66,54 +72,87 @@ function BlockBindingsPanelMenuContent( { fieldsList, attribute, binding } ) {
-						{ Object.entries( fields )
-							.filter(
-								( [ , args ] ) => args?.type === attributeType
-							)
-							.map( ( [ key, args ] ) => (
-								<Menu.RadioItem
-									key={ key }
-									onChange={ () =>
-										updateBlockBindings( {
-											[ attribute ]: {
-												source: name,
-												args: { key },
-											},
-										} )
-									}
-									name={ attribute + '-binding' }
-									value={ key }
-									checked={ key === currentKey }
-								>
+		<Menu placement={ isMobile ? 'bottom-start' : 'left-start' }>
+			{ Object.entries( sources )
+				.filter( ( [ , source ] ) => {
+					// Only show sources that have compatible data for this specific attribute.
+					const sourceDataItems = source.data?.filter(
+						( item ) => item?.type === attributeType
+					);
+					return sourceDataItems && sourceDataItems.length > 0;
+				} )
+				.map( ( [ sourceKey, source ] ) => {
+					if ( source.mode === 'dropdown' ) {
+						return (
+							<Menu key={ sourceKey }>
+								<Menu.SubmenuTriggerItem>
 									<Menu.ItemLabel>
-										{ args?.label }
+										{ source.label }
 									</Menu.ItemLabel>
-									<Menu.ItemHelpText>
-										{ args?.value }
-									</Menu.ItemHelpText>
-								</Menu.RadioItem>
-							) ) }
-					</Menu.Group>
-					{ i !== Object.keys( fieldsList ).length - 1 && (
-						<Menu.Separator />
-					) }
-				</Fragment>
-			) ) }
-		</>
+								</Menu.SubmenuTriggerItem>
+								<Menu.Popover gutter={ 8 }>
+									<Menu.Group>
+										{ source.data
+											?.filter(
+												( item ) =>
+													item?.type === attributeType
+											)
+											.map( ( item ) => (
+												<Menu.RadioItem
+													key={ item.key }
+													onChange={ () => {
+														source.onSelect( {
+															item,
+															updateBlockBindings,
+															attribute,
+														} );
+													} }
+													name={
+														attribute + '-binding'
+													}
+													value={ item.key }
+													checked={
+														item.key === currentKey
+													}
+												>
+													<Menu.ItemLabel>
+														{ item?.label }
+													</Menu.ItemLabel>
+													<Menu.ItemHelpText>
+														{ item?.value }
+													</Menu.ItemHelpText>
+												</Menu.RadioItem>
+											) ) }
+									</Menu.Group>
+								</Menu.Popover>
+							</Menu>
+						);
+					}
+
+					if ( source.mode === 'modal' ) {
+						return (
+							<Menu.Item
+								key={ sourceKey }
+								onClick={ () =>
+									onOpenModal( { attribute, sourceKey } )
+								}
+							>
+								<Menu.ItemLabel>
+									{ source.label }
+								</Menu.ItemLabel>
+							</Menu.Item>
+						);
+					}
+
+					return null;
+				} ) }
+		</Menu>
 	);
 }
 
-function BlockBindingsAttribute( { attribute, binding, fieldsList } ) {
+function BlockBindingsAttribute( { attribute, binding, sources } ) {
```

Comment:
For the sake of information parsimony, could we accept only one `source` here? (I also find it a bit easier to mentally model the correspondence: One `BlockBindingsAttribute` corresponds to one source -- the one that the attribute is bound to.)